src/node_quic_session.cc

```diff
@@ -190,6 +192,493 @@ void QuicSessionConfig::SetQlog(const ngtcp2_qlog_settings& qlog) {
+    AddressToJS(env, remote)
+  };
+  // Grab a shared pointer to this to prevent the QuicSession
+  // from being freed while the MakeCallback is running.
+  BaseObjectPtr<QuicSession> ptr(Session());
+  Session()->MakeCallback(
+      env->quic_on_session_path_validation_function(),
+      arraysize(argv),
+      argv);
+}
+
+void JSQuicSessionListener::OnSessionTicket(int size, SSL_SESSION* session) {
+  Environment* env = Session()->env();
+  HandleScope scope(env->isolate());
+  Context::Scope context_scope(env->context());
+
+  unsigned int session_id_length;
+  const unsigned char* session_id_data =
+      SSL_SESSION_get_id(session, &session_id_length);
+
+  Local<Value> argv[] = {
+    Buffer::Copy(
+        env,
+        reinterpret_cast<const char*>(session_id_data),
+        session_id_length).ToLocalChecked(),
+    v8::Undefined(env->isolate()),
+    v8::Undefined(env->isolate())
+  };
+
+  AllocatedBuffer session_ticket = env->AllocateManaged(size);
+  unsigned char* session_data =
+    reinterpret_cast<unsigned char*>(session_ticket.data());
+  memset(session_data, 0, size);
+  i2d_SSL_SESSION(session, &session_data);
+  if (!session_ticket.empty())
+    argv[1] = session_ticket.ToBuffer().ToLocalChecked();
+
+  if (Session()->IsFlagSet(
+          QuicSession::QUICSESSION_FLAG_HAS_TRANSPORT_PARAMS)) {
+    argv[2] = Buffer::Copy(
+        env,
+        reinterpret_cast<const char*>(&Session()->transport_params_),
+        sizeof(Session()->transport_params_)).ToLocalChecked();
+  }
+  // Grab a shared pointer to this to prevent the QuicSession
+  // from being freed while the MakeCallback is running.
+  BaseObjectPtr<QuicSession> ptr(Session());
+  Session()->MakeCallback(
+      env->quic_on_session_ticket_function(),
+      arraysize(argv), argv);
+
+}
+
+void JSQuicSessionListener::OnSessionSilentClose(
+    bool stateless_reset,
+    QuicError error) {
+  Environment* env = Session()->env();
+  HandleScope scope(env->isolate());
+  Context::Scope context_scope(env->context());
+
+  Local<Value> argv[] = {
+    stateless_reset ? v8::True(env->isolate()) : v8::False(env->isolate()),
+    Number::New(env->isolate(), static_cast<double>(error.code)),
+    Integer::New(env->isolate(), error.family)
+  };
+
+  // Grab a shared pointer to this to prevent the QuicSession
+  // from being freed while the MakeCallback is running.
+  BaseObjectPtr<QuicSession> ptr(Session());
+  Session()->MakeCallback(
+      env->quic_on_session_silent_close_function(), arraysize(argv), argv);
+}
+
+void JSQuicSessionListener::OnVersionNegotiation(
+    uint32_t supported_version,
+    const uint32_t* vers,
+    size_t vcnt) {
+  Environment* env = Session()->env();
+  HandleScope scope(env->isolate());
+  Local<Context> context = env->context();
+  Context::Scope context_scope(context);
+
+  std::vector<Local<Value>> versions;
+  for (size_t n = 0; n < vcnt; n++)
+    versions.push_back(Integer::New(env->isolate(), vers[n]));
```

Comment:
Something along the lines of https://github.com/nodejs/node/pull/30751 might make sense here, i.e. replacing `std::vector` with `MaybeStackBuffer` to avoid heap allocations in the common case

Reply:
Yep, can apply the same to headers handling. good catch.